### PR TITLE
JAMES-2578 Add new Mailet API

### PR DESCRIPTION
**Pull request by blackheaven**
Only since 9dd032f

### CODE REVIEW

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 1]
```diff
@@ -232,9 +245,12 @@ default boolean hasSender() {
      *
      * @return an Iterator (of Strings) over all attribute names
      * @since Mailet API v2.1
+     * @deprecated see {@link #attributeNames()}
      */
     Iterator<String> getAttributeNames();
 
+    Iterator<AttributeName> attributeNames();
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Why not a `Stream<AttributeName>`, which is more convenient for the consumer...

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 2]
```diff
@@ -212,17 +219,23 @@ default boolean hasSender() {
      * @param state the new state of this message
      */
     void setState(String state);
-    
+
+    Stream<Attribute> attributes();
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Add the java doc explaining what this new method does

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 3]
```diff
@@ -212,17 +219,23 @@ default boolean hasSender() {
...
      *
      * @param name the attribute name
      * @return the attribute value, or null if the attribute does not exist
      * @since Mailet API v2.1
+     * @deprecated see {@link #getAttribute(AttributeName)}
      */
+    @Deprecated
     Serializable getAttribute(String name);
     
+    Optional<Attribute> getAttribute(AttributeName name);
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Add the java doc explaining what this new method does

File: mailet/api/src/main/java/org/apache/mailet/Mail.java (hunk 1 above)
Comment (chibenwa) [reply mailet/api/src/main/java/org/apache/mailet/Mail.java]: Add the java doc explaining what this new method does

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 4]
```diff
@@ -251,8 +267,11 @@ default boolean hasSender() {
      *      if there was no such attribute (or if the attribute existed
      *      and its value was null)
      * @since Mailet API v2.1
+     * @deprecated see {@link #removeAttribute(AttributeName)
      */
     Serializable removeAttribute(String name);
+
+    Attribute removeAttribute(AttributeName attributeName);
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Add the java doc explaining what this new method does

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 5]
```diff
@@ -276,9 +295,12 @@ default boolean hasSender() {
      *      or null if there was no such attribute (or if the attribute existed
      *      and its value was null)
      * @since Mailet API v2.1
+     * @deprecated see {@link #setAttribute(Attribute)
      */
     Serializable setAttribute(String name, Serializable object);
 
+    Attribute setAttribute(Attribute attribute);
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Add the java doc explaining what this new method does

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 6]
```diff
@@ -324,4 +346,12 @@ default boolean hasSender() {
      * @since Mailet API v2.3
      */
     void setLastUpdated(Date lastUpdated);
+
+    default Map<AttributeName, Attribute> attributesMap() {
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Add the java doc explaining what this new method does

File: mailet/api/src/main/java/org/apache/mailet/Mail.java (hunk 1 above)
Comment (blackheaven) [reply mailet/api/src/main/java/org/apache/mailet/Mail.java]: Because it's a drop-in replacement, I'll have a look in the main PR to see if we could benefit from it.

**Comment by blackheaven** [comment]
There's no tests for current `*Attribute` methods in `FakeMail`/`MailImpl`, I don't think it worth the investment since it's used everywhere.

File: mailet/api/src/main/java/org/apache/mailet/Mail.java (hunk 1 above)
Comment (blackheaven) [reply mailet/api/src/main/java/org/apache/mailet/Mail.java]: We can benefit from it: I'll change that.

**Comment by chibenwa** [comment]
> I don't think it worth the investment since it's used everywhere.

These tests are not expensive to write, right?

BTW because these methods are used everywhere then these tests are even more needed.

IMO that would be a cool testing enhancement on one core data-model.

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 7]
```diff
@@ -233,7 +236,14 @@ default boolean hasSender() {
      */
     @Deprecated
     Serializable getAttribute(String name);
-    
+
+    /**
+     * Returns the attribute corresponding to an attribute name
+     *
+     * @param name the attribute name
+     * @return the attribute
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: IMO `@return` and `@param` are not adding value here. IMO either we make it more useful or we can drop it - it repeat type and variable name.

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 8]
```diff
@@ -703,7 +707,15 @@ public void dispose() {
      * @since 2.2.0
      */
     public Map<String, Object> getAttributesRaw() {
-        return attributes;
+        return attributes.entrySet().stream()
```
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]:
```
        return ImmutableMap.copyOf(attributes);
```

Seems more straightforward

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 9]
```diff
@@ -749,6 +785,11 @@ public void removeAllAttributes() {
         return attributes.keySet().iterator();
     }
 
+    @Override
+    public Stream<AttributeName> attributeNames() {
+        return attributes.keySet().stream().map(AttributeName::of);
```
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]:
```
        return attributes().map(Attribute::getName);
```

File: mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    @Nested
+    public class EmptyMail {
+        private Mail mail;
+        
+        @BeforeEach
+        public void setUp() {
+            mail = newMail();
+        }
+        
+        public void newMailShouldHaveNoAttributes() {
```
Comment (trantienduchn) [inline mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java]: s/Attributes/Attribute

File: mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            assertThat(mail.attributes()).isEmpty();
+        }
+        
+        @Test
+        public void newMailShouldHaveNoAttributeNames() {
```
Comment (trantienduchn) [inline mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java]: s/AttributeNames/AttributeName

File: mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    }
+    
+    @Nested
+    public class OneAttributeMail {
+        private Mail mail;
+        
+        @BeforeEach
+        public void setUp() {
+            mail = newMail();
+            assertThat(mail.setAttribute(ATTRIBUTE_1)).isNull();
```
Comment (trantienduchn) [inline mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java]: why do we have assertion at setup?

File: mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        
+        @Test
+        public void shouldBeReplacable() {
+            assertThat(mail.setAttribute(ATTRIBUTE_1_BIS)).isEqualTo(ATTRIBUTE_1);
+            assertThat(mail.getAttribute(ATTRIBUTE_NAME_1)).contains(ATTRIBUTE_1_BIS);
+        }
+    }
+    
+    @Nested
+    public class TwoAttributesMail {
```
Comment (trantienduchn) [inline mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java]: this class is missing mail.setAttribute(), mail.getAttribute(), mail.removeAttribute() tests. And my idea is naming it to `MultipleAttributesMail` where you can freely play tests with various number of attributes, not only 2. Could gain more confident.

File: mailet/test/pom.xml [hunk 14]
```diff
@@ -38,6 +38,13 @@
             <groupId>${james.groupId}</groupId>
             <artifactId>apache-mailet-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-mailet-api</artifactId>
+            <version>3.2.0-SNAPSHOT</version>
```
Comment (trantienduchn) [inline mailet/test/pom.xml]: this line seems redundant

File: server/container/core/pom.xml [hunk 15]
```diff
@@ -42,6 +42,13 @@
             <groupId>${james.groupId}</groupId>
             <artifactId>apache-mailet-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-mailet-api</artifactId>
+            <version>3.2.0-SNAPSHOT</version>
```
Comment (trantienduchn) [inline server/container/core/pom.xml]: idem

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 16]
```diff
@@ -703,7 +707,15 @@ public void dispose() {
      * @since 2.2.0
      */
     public Map<String, Object> getAttributesRaw() {
-        return attributes;
+        return attributes.entrySet().stream()
+            .collect(
+                Guavate.toImmutableMap(
```
Comment (trantienduchn) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: Dont use Guavate

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 17]
```diff
@@ -719,26 +731,50 @@ public void dispose() {
      * @param attr Serializable of the entire attributes collection
      * @since 2.2.0
      */
-    public void setAttributesRaw(HashMap<String, Object> attr) {
-        this.attributes = (attr == null) ? new HashMap<>() : attr;
+    public void setAttributesRaw(Map<String, Object> attr) {
+        this.attributes = attr;
```
Comment (trantienduchn) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: why did you remove null check condition? seems dangerous

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 18]
```diff
@@ -703,7 +707,15 @@ public void dispose() {
...
-        return attributes;
+        return attributes.entrySet().stream()
+            .collect(
+                Guavate.toImmutableMap(
+                    x -> x.getKey(),
+                    x -> x.getValue()));
+    }
+
+    public void setAttributes(Map<String, Object> attr) {
+        this.attributes = attr;
```
Comment (trantienduchn) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: no null check?

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 19]
```diff
@@ -719,26 +731,50 @@ public void dispose() {
...
     }
 
     @Override
     public Serializable getAttribute(String key) {
         return (Serializable) attributes.get(key);
     }
 
+    @Override
+    public Optional<Attribute> getAttribute(AttributeName name) {
+        return Optional.ofNullable(attributes.get(name.asString())).map(value -> new Attribute(name, AttributeValue.ofAny(value)));
```
Comment (trantienduchn) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: `name.asString()` is not safe, it could throws NullPointerException. Especially this is a public method, it could be invoked from everywhere and we could not control the null safe guarantee from all of it's clients.

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 20]
```diff
@@ -251,8 +284,20 @@ default boolean hasSender() {
...
      * @since Mailet API v2.1
+     * @deprecated see {@link #removeAttribute(AttributeName)
      */
+    @Deprecated
     Serializable removeAttribute(String name);
+
+    /**
+     * Removes the attribute with the given attribute name from this Mail instance.
+     * 
+     * @return the removed attribute, or null
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Sorry I missed this in my previous reviews, but ain't we supposed to no more design APIs using null?

File: mailet/api/src/main/java/org/apache/mailet/Mail.java [hunk 21]
```diff
@@ -276,9 +321,28 @@ default boolean hasSender() {
...
+     * for Java packages.
+     * The Mailet API specification reserves names matching
+     * <i>org.apache.james.*</i> and <i>org.apache.mailet.*</i>.
+     *
+     * @return the previously existing attribute with the same name,
+     *      or null if there was no such attribute (or if the attribute existed
+     *      and its value was null)
+     * @since Mailet API v3.2
+     */
+    Attribute setAttribute(Attribute attribute);
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Mail.java]: Sorry I missed this in my previous reviews, but ain't we supposed to no more design APIs using null?

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java (hunk 19 above)
Comment (chibenwa) [reply server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: AttributeName is strongly typed and we can assume we can safely call `asString`

Or this object needs strong re-design...

**Comment by chibenwa** [comment]
test this please
